### PR TITLE
feat: Enhance tracing using OpenTelemetry and Grafana Tempo

### PR DESCRIPTION
**Pull request by srodenhuis**
This PR enhances the tracing features in Otomi by:

- Using OpenTelemtry Operator for creating Collectors and Automated Instrumentation
- Introducing the option to use Grafana Tempo as a backend for storing telemetry data using cheap object storage like S3 (by default Tempo is configured to use local Minio)
- Configuring Istio for opentelemetry and using Telemetry resources for envoy access logging
- Configuring Grafana datasources for Tempo to use derived fields to provide a link from the TraceID in the (Loki) logs to the trace in Tempo
- Use node graphs in Grafana to visualize directed graphs or networks
- Configure Nginx ingress controller(s) for opentelemetry
- Automatically configure OpenTelemetry Collector with exporters for Tempo and/or Jaeger

Other changes:
- meshConfig in default is replaced by extensionProviders so users can add their own extensionProviders (like some users already do)

There are still a couple of issues that will need attention later on:

- traces from Nginx and Istio need additional labels for the trace2logs feature
- Resources for the opentelemetry sidecar in the helm chart can not be configured. I changed the upstream chart to set limits. This does solve the issue where nginx consumes around 2 CPU when opentelemtry is enabled

If you have any questions, let me know.

### CODE REVIEW

File: charts/otomi-api/templates/deployment.yaml [hunk 1]
```diff
@@ -40,27 +40,21 @@ spec:
...
-            failureThreshold: 3
-            httpGet:
-              path: /live
-              port: 9000
-            periodSeconds: 10
           ports:
             - name: http
               containerPort: 8080
               protocol: TCP
+          livenessProbe:
```
Comment (j-zimnowoda) [inline charts/otomi-api/templates/deployment.yaml:47]: Current probes are fine. Please restore them

File: charts/team-ns/templates/_ingress.tpl [hunk 2]
```diff
@@ -30,9 +30,8 @@ extensions/v1
 {{- define "ingress" -}}
 {{- $ := . }}
 {{- $v := .dot.Values }}
-{{- $appsDomain := printf "apps-%s.%s" $v.teamId $v.domain }}
+{{- $appsDomain := printf "apps.%s" $v.domain }}
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_ingress.tpl:33]: Please do not overwrite remote changes while merging with main.

File: charts/team-ns/templates/_helpers.tpl [hunk 3]
```diff
@@ -42,17 +42,12 @@ otomi.io/team: {{ .Values.teamId }}
 {{- define "service.domain" -}}
 {{- $v := .dot.Values }}
 {{- $isApps := and .s.isCore (not (or .s.ownHost .s.isShared)) }}
-{{- if and $isApps (not .vs) -}}
-  {{- if eq $v.teamId "admin" -}}
-    {{- printf "apps.%s" $v.domain -}}
-  {{- else -}}
-    {{- printf "apps-%s.%s" $v.teamId $v.domain -}}
-  {{- end -}}
+{{- if and $isApps (not .vs) }}
```
Comment (j-zimnowoda) [inline charts/team-ns/templates/_helpers.tpl:45]: Please do not overwrite remote changes while merging with main.

File: charts/tempo/templates/ingester/poddisruptionbudget-ingester.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {{- if gt (int .Values.ingester.replicas) 1 }}
 {{ $dict := dict "ctx" . "component" "ingester" "memberlist" true }}
-apiVersion: {{ include "tempo.pdb.apiVersion" . }}
```
Comment (j-zimnowoda) [inline charts/tempo/templates/ingester/poddisruptionbudget-ingester.yaml:3]: is community chart incorrect

File: charts/tempo/values.yaml [hunk 5]
```diff
@@ -783,7 +783,7 @@ traces:
       receiverConfig: {}
   zipkin:
     # -- Enable Tempo to ingest Zipkin traces
-    enabled: false
```
Comment (j-zimnowoda) [inline charts/tempo/values.yaml:786]: Let's change this values through the values/**.gotemplate file.

File: helmfile.d/helmfile-60.teams.yaml [hunk 6]
```diff
@@ -26,9 +26,9 @@ bases:
 releases:
 {{- range $teamId, $team := omit $tc "admin" }} 
   {{- $teamServices := ($team | get "services" list) }}
-  {{- $domain := ($v.cluster | get "domainSuffix" nil) }}
+  {{- $domain := printf "team-%s.%s" $teamId ($v.cluster | get "domainSuffix" nil) }}
```
Comment (j-zimnowoda) [inline helmfile.d/helmfile-60.teams.yaml:29]: Please do not overwrite remote changes while merging with main.

File: helmfile.d/helmfile-60.teams.yaml [hunk 7]
```diff
@@ -26,9 +26,9 @@ bases:
...
   {{- $appsDomain := printf "apps.%s" $domain }}
-  {{- $grafanaDomain := printf "grafana-%s.%s" $teamId $domain }}
+  {{- $grafanaDomain := printf "grafana.%s" $domain }}
```
Comment (j-zimnowoda) [inline helmfile.d/helmfile-60.teams.yaml:31]: Please do not overwrite remote changes while merging with main.

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]

File: helmfile.d/helmfile-91.artifacts.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+---
+bases:
+  - snippets/derived.gotmpl
+---
+{{ readFile "snippets/templates.gotmpl" }}
+{{- $v := .Values }}
+{{- $a := $v.apps }}
+
+releases:
+  - name: otel-operator-artifacts
```
Comment (j-zimnowoda) [inline helmfile.d/helmfile-91.artifacts.yaml:15]: This release is already defined in the `helmfile.d/helmfile-90.artifacts.yaml` file

File: helmfile.d/snippets/certificate.gotmpl [hunk 9]
```diff
@@ -0,0 +1,21 @@
+apiVersion: cert-manager.io/v1
```
Comment (j-zimnowoda) [inline helmfile.d/snippets/certificate.gotmpl:1]: Please do not overwrite remote changes while merging with main.

File: helmfile.d/snippets/domains.gotmpl [hunk 10]
```diff
@@ -19,17 +19,16 @@
   {{- $_ := set $tc "admin" dict }}
 {{- end }}
 {{- range $teamId, $team := $tc }}
-  {{- $baseDomain := $v.cluster.domainSuffix }}
```
Comment (j-zimnowoda) [inline helmfile.d/snippets/domains.gotmpl:22]: Revert this change please

File: src/cmd/validate-templates.ts [hunk 11]
```diff
@@ -6,7 +6,7 @@ import { terminal } from 'src/common/debug'
 import { hfTemplate } from 'src/common/hf'
```
Comment (j-zimnowoda) [inline src/cmd/validate-templates.ts:6]: Please revert changes in this file, as they are not relevant.

File: upgrades/pre/upgrade-0-25-0.sh [hunk 12]
```diff
@@ -2,6 +2,6 @@
 
 set -eu
 
-kubectl get certificate -n istio-system -o name | while read -r cert; do kubectl annotate "$cert" -n istio-system helm.sh/resource-policy='keep' deprecated=true; done
```
Comment (j-zimnowoda) [inline upgrades/pre/upgrade-0-25-0.sh:5]: Please revert changes in this file, as they are not relevant.

File: values-schema.yaml [hunk 13]
```diff
@@ -1569,13 +1578,6 @@ properties:
             $ref: '#/definitions/rawValues'
           resources:
             $ref: '#/definitions/resources'
-          byoWildcardCert:
```
Comment (j-zimnowoda) [inline values-schema.yaml:1572]: Why is it removed ?

File: values-schema.yaml [hunk 14]
```diff
@@ -2199,9 +2220,11 @@ properties:
                 properties:
                   resources:
                     $ref: '#/definitions/resources'
-          meshConfig:
```
Comment (j-zimnowoda) [inline values-schema.yaml:2202]: Please provide the values-changes, since you remove the property

File: values/cert-manager/cert-manager-raw.gotmpl [hunk 15]
```diff
@@ -2,7 +2,10 @@
 {{- $cm := index $v.apps "cert-manager" }}
 {{- $dns := $v | get "dns" dict }}
 {{- $p := $dns | get "provider" dict }}
+{{- $certTpl := readFile "../../helmfile.d/snippets/certificate.gotmpl" }}
```
Comment (j-zimnowoda) [inline values/cert-manager/cert-manager-raw.gotmpl:5]: again, incorrect merge

File: values/ingress-nginx/ingress-nginx.gotmpl [hunk 16]
```diff
@@ -43,8 +43,9 @@ controller:
     {{- end }}
   podAnnotations:
     # sidecar.istio.io/inject: {{ if $v.otomi.hasCloudLB }}"true"{{ else }}"false"{{ end }}
-    policy.otomi.io/ignore: psp-privileged
-
+    policy.otomi.io/ignore: psp-privileged,psp-allowed-users
```
Comment (j-zimnowoda) [inline values/ingress-nginx/ingress-nginx.gotmpl:46]: we should not run nginx-ingress controller as root

File: values/otel-operator/otel-operator-raw.gotmpl [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+          otlp:
+            endpoint: tempo-distributor.tempo.svc.cluster.local:4317
+            sending_queue:
+              enabled: true
+              num_consumers: 100
+              queue_size: 10000
+            retry_on_failure:
+              enabled: true
+            tls:
+              insecure: true
```
Comment (j-zimnowoda) [inline values/otel-operator/otel-operator-raw.gotmpl:36]: Can we change this flag to `insecure: false`

File: values/otel-operator/otel-operator-raw.gotmpl [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+          {{- end }}
+          {{- if $v.apps.jaeger.enabled }}
+          jaeger:
+            endpoint: jaeger-operator-jaeger-collector.jaeger.svc:14250
+            tls:
+              insecure: true
```
Comment (j-zimnowoda) [inline values/otel-operator/otel-operator-raw.gotmpl:42]: Can we change this flag to insecure: false

File: values/otomi-api/otomi-api.gotmpl [hunk 19]
```diff
@@ -39,7 +39,7 @@ secrets:
   {{- $sopsEnv | nindent 2 }}
   
 env:
-  DEBUG: 'otomi:*,-otomi:authz,-otomi:repo'
```
Comment (j-zimnowoda) [inline values/otomi-api/otomi-api.gotmpl:42]: Please revert changes in this file, as they are not relevant.

File: values/tempo/tempo.gotmpl [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+  resources: {{- $t.resources.memcached | toYaml | nindent 4 }}
+
+metricsGenerator:
+  enabled: true
+  resources: {{- $t.resources.metricsGenerator | toYaml | nindent 4 }}
+  config:
+    storage:
+      path: /var/tempo/wal
+      wal:
+      remote_write_flush_deadline: 1m
```
Comment (j-zimnowoda) [inline values/tempo/tempo.gotmpl:64]: it looks like indentation error

File: values/prometheus-operator/prometheus-operator.gotmpl [hunk 21]
```diff
@@ -266,13 +269,62 @@ grafana:
...
         basicAuthPassword: {{ $v.apps.loki.adminPassword }}
+    {{- if $c.tempo.enabled }}
+      jsonData:
+        derivedFields:
+          - datasourceName: Tempo
+            matcherRegex: "traceID=00-([^\\-]+)-"
+            name: TraceID
+            url: "$${__value.raw}"
+            datasourceUid: tempo
+    - name: Tempo
```
Comment (j-zimnowoda) [inline values/prometheus-operator/prometheus-operator.gotmpl:288]: It looks like there is a big overlap with the configuration that you added to the `helmfile.d/helmfile-60.teams.yaml` file. Why is it necessary?

File: helmfile.d/helmfile-91.artifacts.yaml (hunk 8 above)
Comment (srodenhuis) [reply helmfile.d/helmfile-91.artifacts.yaml:15]: indeed

File: values/otel-operator/otel-operator-raw.gotmpl (hunk 17 above)
Comment (srodenhuis) [reply values/otel-operator/otel-operator-raw.gotmpl:36]: no. but is only in cluster

File: values/otel-operator/otel-operator-raw.gotmpl (hunk 18 above)
Comment (srodenhuis) [reply values/otel-operator/otel-operator-raw.gotmpl:42]: no

File: values/prometheus-operator/prometheus-operator.gotmpl (hunk 21 above)
Comment (srodenhuis) [reply values/prometheus-operator/prometheus-operator.gotmpl:288]: Because there is more than 1 grafana instance (platform grafana and team's grafana). Both have separate configs

File: values/tempo/tempo.gotmpl (hunk 20 above)
Comment (srodenhuis) [reply values/tempo/tempo.gotmpl:64]: yes, looks like a wrong indent, but this is not the case ;-)

File: charts/tempo/templates/ingester/poddisruptionbudget-ingester.yaml (hunk 4 above)
Comment (srodenhuis) [reply charts/tempo/templates/ingester/poddisruptionbudget-ingester.yaml]: I don't know, but without changing api version, the chart does not pass the validate-templates

File: values/tempo/tempo.gotmpl [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+      enabled: true
+      include_all_attributes: true
+      filter_by_status_error: true
+
+compactor:
+  resources: {{- $t.resources.compactor | toYaml | nindent 4 }}
+
+querier:
+  resources: {{- $t.resources.querier | toYaml | nindent 4 }}
+#   podAnnotations:
```
Comment (j-zimnowoda) [inline values/tempo/tempo.gotmpl:36]: Could you remove unused config?

File: values/tempo/tempo.gotmpl (hunk 20 above)
Comment (j-zimnowoda) [reply values/tempo/tempo.gotmpl:64]: hmm then what is the value for `wal` 
Can set wal to null ?

File: values/istio-operator/istio-operator-raw.gotmpl [hunk 23]
```diff
@@ -149,11 +149,27 @@ resources:
...
+          defaultConfig:
+            gatewayTopology:
+              numTrustedProxies: 1
+          enableAutoMtls: true
+          extensionProviders:
+          - EnvoyFileAccessLogProvider:
+            name: envoy-access-logs
+            path: /dev/stdout
+          {{- if $i.tracing.enabled }}
+            logFormat: |
```
Comment (j-zimnowoda) [inline values/istio-operator/istio-operator-raw.gotmpl:163]: Depending on traceSampling not every log will have the `TRACEPARENT`.
Have you tested that scenario?

File: helmfile.d/snippets/derived.gotmpl [hunk 24]
```diff
@@ -64,11 +64,11 @@ environments:
             enabled: {{ and (eq $provider "aws") $v.otomi.hasCloudLB }}
           aws-ebs-csi-driver:
             enabled: false
+          {{- if and (eq $issuer "letsencrypt") (not (hasKey $cm "stage")) }}
```
Comment (j-zimnowoda) [inline helmfile.d/snippets/derived.gotmpl:67]: Was the before logic incorrect ?

File: apps.yaml [hunk 25]
```diff
@@ -168,6 +168,7 @@ appsInfo:
     integration: Otomi has security best practices built in, and is designed for intrusion. Istio is used by Otomi as a service mesh to deliver mTLS enforcement for all traffic that is deemed compromisable, egress control to force teams to choose explicit egress endpoints, and advanced routing capabilities such as weight based load balancing (A/B or blue/green testing). Istio is part of the core of Otomi and can not be disabled.
   jaeger:
```
Comment (j-zimnowoda) [inline apps.yaml:169]: Why do we still need Jaeger if traces can be visible in Grafana ?

File: values-schema.yaml [hunk 26]
```diff
@@ -2930,6 +3027,37 @@ properties:
...
+                type: integer
+                default: 1
+          resources:
+            additionalProperties: false
+            properties:
+              collector:
+                $ref: '#/definitions/resources'
+              manager:
+                $ref: '#/definitions/resources'
+              kubeRBACProxy:
```
Comment (j-zimnowoda) [inline values-schema.yaml:3059]: What is the kubeRBACProxy ?

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
I added the `tests/integration/monitoring-with-team.yaml` , 

could you define workloads that are instrumented to forward Headers, so we can easily deploy setup to evaluate the new tracing stack?

File: values/tempo/tempo.gotmpl (hunk 20 above)
Comment (srodenhuis) [reply values/tempo/tempo.gotmpl:64]: see line 321: https://github.com/redkubes/otomi-core/blob/sr/feat/tempo/charts/tempo/values.yaml

File: values/istio-operator/istio-operator-raw.gotmpl (hunk 23 above)
Comment (srodenhuis) [reply values/istio-operator/istio-operator-raw.gotmpl:163]: If tracing is not enabled, envoy will use the default log formatting (without TRACEPARENT)

File: values/istio-operator/istio-operator-raw.gotmpl [hunk 27]
```diff
@@ -149,11 +149,27 @@ resources:
                 memory: 80Mi
             {{- end }}
             autoInject: enabled
-          tracer:
-            zipkin:
-              address: jaeger-operator-jaeger-collector.jaeger:9411 
           useMCP: false
-        meshConfig: {{- $i.meshConfig | toYaml | nindent 10 }}
+        meshConfig:
```
[the remark this answers is not in the document]
Comment (srodenhuis) [reply values/istio-operator/istio-operator-raw.gotmpl:153]: No, this is the default path and will also be used if an extention provider is used

File: apps.yaml (hunk 25 above)
Comment (srodenhuis) [reply apps.yaml:169]: No. Only when Tempo is enabled, grafana datasources are configured. For now you can use both (Jaeger and Tempo/Grafana), but we could (later) make the user choose and if Jaeger is used, configure Grafana to add a link to a traceid in the logs (Loki) and open the trace in Jaeger

**Review by j-zimnowoda — APPROVED** [review]

**Comment by j-zimnowoda** [comment]
Impressive work @srodenhuis !
It was a pleasure to work on this project.

File: values-schema.yaml (hunk 14 above)
Comment (j-zimnowoda) [reply values-schema.yaml:2240]: done
